internal/policy/container/has_modified_files.go

```diff
@@ -45,118 +48,155 @@ func (p *HasModifiedFilesCheck) Validate(ctx context.Context, imgRef image.Image
 // installed via packages from the container image,and the list of files (packageFilesRef.LayerFiles)
 // modified/added via layers in the image.
 func (p *HasModifiedFilesCheck) getDataToValidate(ctx context.Context, imgRef image.ImageReference) (*packageFilesRef, error) {
-	logger := logr.FromContextOrDiscard(ctx)
-
-	// Get a list of packages from the RPM database. This avoids having to rely on
-	// rpm, dnf, yum, etc. being installed in the image.
-	pkgList, err := rpm.GetPackageList(ctx, imgRef.ImageFSPath)
+	fs := afero.NewOsFs()
+	layerDir, err := afero.TempDir(fs, "", "rpm-layers-*")
+	defer func() {
+		_ = fs.RemoveAll(layerDir)
+	}()
 	if err != nil {
-		return nil, fmt.Errorf("could not get rpm list: %w", err)
+		return nil, err
 	}
 
-	packageFiles, err := p.getInstalledFilesFor(pkgList)
-	if err != nil {
-		return nil, fmt.Errorf("could not list installed files: %w", err)
+	if imgRef.ImageInfo == nil {
+		return nil, fmt.Errorf("image reference invalid")
 	}
 
 	layers, err := imgRef.ImageInfo.Layers()
 	if err != nil {
-		return nil, fmt.Errorf("could not get image layers: %w", err)
+		return nil, err
 	}
 
-	files := make([][]string, 0, len(layers))
+	layerIDs := make([]string, 0, len(layers))
+
+	files := make(map[string][]string, len(layers))
+	layerPackages := make(map[string]map[string]struct{}, len(layers))
+	layerPackageFiles := make(map[string]map[string]string, len(layers))
 
-	// Uncompress each layer and build a slice containing the files
-	// modified by each layer.
 	for _, layer := range layers {
-		r, err := layer.Uncompressed()
+		layerID, err := layer.DiffID()
+		if err != nil {
+			return nil, fmt.Errorf("unable to retrieve diff id for layer: %w", err)
+		}
+
+		layerDir := filepath.Join(layerDir, layerID.String())
+		err = fs.Mkdir(layerDir, 0o755)
 		if err != nil {
-			return nil, fmt.Errorf("could not extract layers: %w", err)
+			return nil, fmt.Errorf("could not create layer directory: %w", err)
 		}
-		pathChan := make(chan string)
-
-		var wg sync.WaitGroup
-		wg.Add(1)
-		go func() {
-			// For each path in the pathChan, add it to the layer's
-			// list of files.
-			layerFiles := make([]string, 0)
-			for path := range pathChan {
-				layerFiles = append(layerFiles, path)
+
+		if err := func() error {
```

Comment:
Why an anonymous function here, vs. a standalone?

Reply:
It's a closure. I could pull it out into it's own standalone function though.